core/src/price_estimation/oneinch.rs

```diff
@@ -0,0 +1,414 @@
+mod api;
+
+use super::{PriceSource, TokenData};
+use crate::http::HttpFactory;
+use crate::models::TokenId;
+use anyhow::{anyhow, Context, Result};
+use api::{OneinchApi, OneinchHttpApi};
+use futures::{
+    future::{self, BoxFuture, FutureExt as _},
+    lock::Mutex,
+};
+use std::collections::HashMap;
+
+struct ApiTokens {
+    // Maps uppercase Token::symbol to Token.
+    // This is cached in the struct because we don't expect it to change often.
+    tokens: HashMap<String, api::Token>,
+    stable_coin: api::Token,
+}
+
+pub struct OneinchClient<Api> {
+    api: Api,
+    /// Lazily retrieved the first time it is needed when `get_prices` is
+    /// called. We don't want to use the network in `new`.
+    api_tokens: Mutex<Option<ApiTokens>>,
+    tokens: TokenData,
+}
+
+impl OneinchClient<OneinchHttpApi> {
+    /// Create a OneinchClient using OneinchHttpApi as the api implementation.
+    pub fn new(http_factory: &HttpFactory, tokens: TokenData) -> Result<Self> {
+        let api = OneinchHttpApi::new(http_factory)?;
+        Ok(Self::with_api_and_tokens(api, tokens))
+    }
+}
+
+impl<Api> OneinchClient<Api>
+where
+    Api: OneinchApi,
+{
+    pub fn with_api_and_tokens(api: Api, tokens: TokenData) -> Self {
+        Self {
+            api,
+            api_tokens: Mutex::new(None),
+            tokens,
+        }
+    }
+
+    async fn create_api_tokens(&self) -> Result<ApiTokens> {
+        let tokens = self.api.get_token_list().await?;
+        let mut tokens: HashMap<String, api::Token> = tokens
+            .into_iter()
+            .map(|token| (token.symbol.to_uppercase(), token))
+            .collect();
+
+        // 1inch does track OWL, but all other price sources use some
+        // USD-equivalent token to compute the price of other tokens.
+        // The price of OWL is far from 1 USD at the time of writing
```

Comment:
That's not the case for neither Dex.ag nor Kraken, so if we used owl we'd have only one price source left. I agree that in principle it would be better to use OWL however.

Reply:
I think it should be DAI for Kraken and Dex.AG but OWL for the exchanges that support it. We can pass in the stablecoin token into the constructor.